Add proxy feature

cameronblandford opened the pull request:
This PR contains several bugs which are currently being worked out, most notably the inability to save relationships due to, presumably, an adapter issue.

Resolves https://github.com/OA-PASS/pass-ember/issues/740 https://github.com/OA-PASS/pass-ember/issues/742 https://github.com/OA-PASS/pass-ember/issues/741

karenhanson commented:
Copying issue list from previous PR: 
resolves #730, #728, #729, #726, #727, #732,

karenhanson commented:
I tested this generally - I'm not clear on what is in scope for this PR, so I may have listed some things that are not yet implemented. Also, a few things are possible enhancements, and may require separate issues. I have tried to split these according to issue#. Please let me know if you need better steps to replicate, or screenshots.

## #726 - add button fields to assign the submitter
* I was able to link a user by entering email, that user became the submitter and I the preparer - that worked.
* I logged in as faculty1, I put in a random email that is not in the database. It didn't appear to do anything when I hit "Assign Submitter". When I opened the console log it said "No wesults fouwnd! Sowwy!!! uwu (●´ω｀●)". I entered article information and continued through anyway, and was able to complete a submission with me as preparer and no submitter. Need to indicate to the user when there is no submitter assigned and prevent them from going next if they have not either linked a submitter OR entered an email/name if they have selected "Someone else".
* [enhancement] I would like some kind of visual cue that the user is connected - once I put in a correct email address, a green popup was shown then disappears. I think there needs to be some refinement here - but perhaps this should wait until we have the indexer support in place.
* [enhancement] I wonder if the prepare on behalf of radio button bar should be a little less muted - maybe we should force them to pick one or the other? I think it could be easily missed.

## #729 - show proxy submitter names
* if you put in the name/email of the user that doesn't exist yet, on the submission summary page, it says "submitted on behalf of ()" it should use the name entered.
![nobehalfofname](https://user-images.githubusercontent.com/4070836/46560133-955d5d00-c8c0-11e8-9a69-74ad28228159.png)
* Go in as submitter after preparer requested changes. There are 2 "Request More Changes" buttons - one should be something like "Cancel Submission"
![image](https://user-images.githubusercontent.com/4070836/46560267-09980080-c8c1-11e8-9f22-71be35899143.png)

* If I go in to approve a submission, on Submission Details, I put in a comment and hit "Request more changes". Nothing happens on the webpage. The dev console says "Requested more changes from preparer" but the submission details page does not change and the comment doesn't appear to be saved. When looking in the database, the `SubmissionEvent` has been created but there is no `SubmissionEvent.submission` value. Same happens with the cancel and approve buttons - an unlinked record is created in SubmissionEvents. It does, however, update the `Submission.submissionStatus`, so if you hit "approve", next time you reload Submission Details page it shows it as `submitted`.

## #730, #732 - commenting
* In the comments section on the submission details page, the user name is very small compared to other text. Also, the link on the commenter name doesn't seem to be helpful and there is a sideways ellipsis that looks like a link that can be removed(?)
![commentsidewayselipsis](https://user-images.githubusercontent.com/4070836/46560464-db66f080-c8c1-11e8-8284-582a7fa92f51.png)
* [enhancement/bug?] if there are no comments, I think the comment box seen during the workflow steps should be hidden. It looks odd appearing on every page when there are no comments
![image](https://user-images.githubusercontent.com/4070836/46561078-094d3480-c8c4-11e8-92a0-406880f93e81.png)
* As far as I can tell the ability to edit a submission is not yet implemented so I cannot test whether the comments work throughout the workflow. Also because of a bug with saving SubmissionEvent described in the previous section, I could not test display of multiple comments - will need to test these later.

## #727 - show grants for user you are submitted on behalf of
* If I do a submission for myself, grants list works fine. If I prepare on behalf of someone, I cannot pick a grant. No matter what I enter the dropdown is empty. After talking to @htpvu , one of 2 things should happen here when preparing on behalf of someone. If a person that exists in Fedora is linked to the Submission, their grants should appear as they would if you were submitting for yourself. If the person is not in the database and instead the preparer enters an email/name, then no grants should appear because the person is not a PI for any. This eliminates the need for a grants search index function

## #728 - hide J10P deposit agreement
* deposit agreement is hidden when I am the preparer, but now when I am the submitter, I am seeing additional author metadata on the jscholarship agreement form. will that go away when we start using agreementText instead of formSchema?
![image](https://user-images.githubusercontent.com/4070836/46561115-313c9800-c8c4-11e8-9418-16b4299b280a.png)

* If I go in as submitter, I don't yet see a place where I can sign agreement text - is that not implemented yet?

## #733 - show submission statistics on dashboard
* dashboard is showing I have 10 submissions awaiting approval by submitter, but they are all prepared by me and most are complete. Looks like it might not be using "AND" to filter search?
* I have "0" submissions awaiting approval though there should be 1.

## general
* when I hit submit, the spinner while uploading is now off to the side
![loadingspinnertoside](https://user-images.githubusercontent.com/4070836/46560526-12d59d00-c8c2-11e8-8439-9b86f0225bda.png)
* once I have submitted a submission prepared for someone else, it does not appear in the preparer's submissions list. The submitter AND the preparer should be able to see the submission.

cameronblandford commented:
## Update on local branch progress

(I'll push everything once all the updates are made so you can test them all at the same time, which, thank you for testing the PR, half-baked as it is!)

* #726: bullet 2 is fixed - you now can't proceed without a user or email attached to the submission if you choose to submit as a proxy, and the accidentally committed cutesy debug message has been replaced with an error-style toast (temporary popup) explaining that no user was found.
* #729: bullet 1 and bullet 2 are fixed.
* #730, #732: typography on commenter name has been fixed to be a more standard size, with the options button removed. Comments window is now hidden when no comments are present.
* general: loader now displays centered, as expected.

karenhanson commented:
Having a few issues that are making it difficult to test - can you let me know if you can replicate them @htpvu or @cameronblandford?  I am using the same image versions as are in the `.shib` folder.

* If I do a Submission for myself, when I enter a title/journal and hit next, I get a "something went wrong" error.  The console shows `POST https://pass.local/es/ 400 (Bad Request)`. The same thing happens if I enter a new email/name for someone that doesn't exist.
* If I do a Submission for someone else, I can get through to the end of and when I request approval, I get an empty red error box top right. Console says `Failed to load resource: the server responded with a status of 400 (400)`. When I looked at the server log I see this:
```
fcrepo           | 21:25:26.334 [          http-nio-8080-exec-3] INFO (FedoraLdp) GET resource 'journals/30/13/bb/54/3013bb54-14ac-4683-a02a-9ba163f64fb8'
proxy            | httpd-combined 172.23.0.1 - - [12/Oct/2018:21:25:26 +0000] "POST /fcrepo/rest/submissions HTTP/1.1" 400 34 "https://pass.local/app/submissions/new" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36"
fcrepo           | 21:25:26.356 [           ajp-nio-8009-exec-3] WARN (DeserializationFilter) Bad request
fcrepo           | org.dataconservancy.fcrepo.jsonld.BadRequestException: Unknown attribute submissionStatus
fcrepo           |      at org.dataconservancy.fcrepo.jsonld.JsonldNtriplesTranslator.verify(JsonldNtriplesTranslator.java:121)
fcrepo           |      at org.dataconservancy.fcrepo.jsonld.JsonldNtriplesTranslator.translate(JsonldNtriplesTranslator.java:76)
fcrepo           |      at org.dataconservancy.fcrepo.jsonld.deserialize.DeserializationWrapper.<init>(DeserializationWrapper.java:67)
fcrepo           |      at org.dataconservancy.fcrepo.jsonld.deserialize.DeserializationFilter.doFilter(DeserializationFilter.java:94)
fcrepo           |      at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
fcrepo           |      at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
fcrepo           |      at org.dataconservancy.fcrepo.jsonld.compact.CompactionFilter.doFilter(CompactionFilter.java:117)
fcrepo           |      at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
fcrepo           |      at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
fcrepo           |      at org.apache.catalina.filters.RequestDumperFilter.doFilter(RequestDumperFilter.java:203)
fcrepo           |      at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
fcrepo           |      at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
fcrepo           |      at org.apache.catalina.filters.CorsFilter.handleSimpleCORS(CorsFilter.java:312)
fcrepo           |      at org.apache.catalina.filters.CorsFilter.doFilter(CorsFilter.java:161)
fcrepo           |      at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
fcrepo           |      at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
fcrepo           |      at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
fcrepo           |      at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
fcrepo           |      at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:613)
fcrepo           |      at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
fcrepo           |      at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
fcrepo           |      at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:650)
fcrepo           |      at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
fcrepo           |      at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
fcrepo           |      at org.apache.coyote.ajp.AjpProcessor.service(AjpProcessor.java:479)
fcrepo           |      at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
fcrepo           |      at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:790)
fcrepo           |      at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1468)
fcrepo           |      at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
fcrepo           |      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
fcrepo           |      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
fcrepo           |      at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
fcrepo           |      at java.lang.Thread.run(Thread.java:748)
fcrepo           | 21:25:26.518 [          http-nio-8080-exec-5] INFO (FedoraLdp) GET resource 'publications/a8/e3/d9/90/a8e3d990-70c1-46e1-ba6e-f1b3c426a25f'
```

Things that I could see were fixed:
* If I select a submitter by email, I can see their grants
* the loader icon is now centralized when you hit submit

Things that didn't appear to be fixed:
* I can hit next on Basics tab if select "Someone Else" and don't find a matching user or provide a new email/user
* The font was fixed on user name above comments box on Submission Details, but the link on email should be a "mailto:" link, currently it links to a Fedora record, which could be dangerous.

Since I was unable to complete a Submission, I could not test anything else.

cameronblandford commented on app/controllers/submissions/new.js:
```diff
@@ -33,65 +53,136 @@ export default Controller.extend({
        * Where calling `obj.set('prop', obj2.get('id'))` will not set the relationship
        */
       const pub = this.get('model.publication');
-      sub.set('aggregatedDepositStatus', 'not-started');
```
I was under the impression that the deposit service handled these statuses, which is why I removed them. Is that not true?

cameronblandford commented:
@karenhanson have you made sure that all of your containers are up to date and that you are in fact losing the latest ones? I can't replicate your first issue. As for the second, you have to log in as a user for it to be available in the database. I'm not sure what the cause of that is, but my guess is that the auth credentials for the users exist without the user objects being created yet. So you'd have to log in in an incognito tab as the user you want to assign to the submitter role before they're able to show up.

cameronblandford commented on app/controllers/submissions/new.js:
```diff
@@ -33,65 +53,136 @@ export default Controller.extend({
        */
       const pub = this.get('model.publication');
-      sub.set('aggregatedDepositStatus', 'not-started');
+      sub.set('submissionStatus', 'not-started');
```
Good catch, thanks!

karenhanson commented:
> but isn't this a security vulnerability in general if that information is considered dangerous but is available to any logged in user on the repository?
The authz image will be implemented at the same time and that prevents inappropriate edits, so not really dangerous. Probably "confusing" would be a better word.

karenhanson commented:
The issues I was having were a result of an environment variable in `pass-docker`. I have created an issue for updating the Ember image in `pass-docker` and made note of the required env variable change.
https://github.com/OA-PASS/pass-docker/issues/133

karenhanson commented:
@cameronblandford @htpvu  OK, this round of testing was much more productive now that i have my envs in order! Here is the bullet list of my findings, please let me know if you need more information:

### Workflow - basics step
Note: These can be skipped until user search integration is done, but I wanted to make note of them as a reminder to check if they are still relevant later.
* If I go back to the basics tab after after linking a user by email, the user field is empty. Once we have improved user search, I think we need to display some info about the linked User and have an option to change or clear it.
*  I can select my own email address to submit on behalf of, it then allows me to add a comment in the last step, but does not make me the preparer so no harm done. When we implement this for real, perhaps we should exclude the current user's record from the search results.

### Workflow - grants step
* If I start a new submission from the `/grants/` list then decide to submit on behalf of someone, my grant is still assigned, so it is possible to mix my grant with those of the designated submitter.
* When I add grants from the dropdown, they used to disappear from the list after you add them. Can that functionality be put back?
* When I choose to create a _new_ user, I enter a name, email and click next. An error occurs, which seems to be caused by being unable to retrieve Grants. Grants should be an empty list in this scenario so no lookup is necessary.

### Workflow - files step
* If I edit a submission (as either preparer or submitter), I see the existing files list in the workflow, but it will not let me proceed unless I upload another file.

### Workflow - review step:
* If I do 2 submissions back to back, the comment from the previous submission is still cached on the submission review page
* If I edit an existing submission either as a preparer or submitter and it has a linked repository, I am asked to visit that repository again.
* If I edit a submission prepared by someone else, on the submission review page of the workflow, the comments section talks about "This submission is prepared on behalf of Ed User. Once Request approval is clicked..." I think the text above the comments box needs to change based on who is viewing it and the current status.
* enhancement-ish: If the submitter requests edits, and I go in as the preparer, edit the submission, next through the workflow, on the submission details page, I would expect to see the previous comments above the comment box, but there are none.

### Submission Details page:
* When I get to review section of `/submissions/new`, the mailto link now works. If I then go to `/submissions/{uri}` after the Submission record is created, the preparer's User name in the comment section is still linking to the submission details page. It should be formatted the same way as seen on `/submissions/new` review page.
* enhancement-ish: If I cancel or request changes to a submission prepared by someone else, the Submission.submissionStatus is correct, but there is no indication of the current status on the Submission Details page. First, I propose changing the text for the Status field in the Repositories box to "Not submitted" instead of "Waiting to be submitted" so that it doesn't look strange when a submission is cancelled. Second, now that there are various pre-submission statuses, I think we need to display the submissionStatus on the Submission Details page, probably near the top so that you can immediately see "ok, now my submission is cancelled", and also probably with some colors to make it stand out.

### Submissions list page
* Log in to look at submission awaiting approval, request more changes, in `/submissions/` list, the status is now "Edits required". The phrasing in the tooltip is directed at the preparer but I am the submitter. Perhaps it could be more generic "The authorized submitter has requested some changes to the submission" - i.e. don't say "you"
* I think yous aid this is a separate PR/issue that hasn't been addressed yet, but the statuses on the `/submissions/` page need to be cleaned up. They should now just print pretty much what the database says for submissionStatus. e.g. manuscript-expect -> "Manuscript expected" etc. The status button can just link to the submission details page as it does with `See details`. We will need to rethink the color scheme and tooltip text. For example, currently when I request approval, the status is `approval-requested`, the display should say "Approval requested", currently it says "See details".

### Dashboard
* enhancement: dashboard is now correctly showing 1 submission waiting, but perhaps it can link to /submissions/ page - I wanted to be able to click on it.

karenhanson commented:
@cameronblandford I found another bug on the Submission Details page:
* If I prepare on behalf of someone, then log in as that user and hit submit. The details page shows this, where presumably it show a person name for submitter, and maybe preparer name(s) too where there is one:
![image](https://user-images.githubusercontent.com/4070836/47182881-a3fe3800-d2f4-11e8-8c91-25a2516ca7b7.png)

karenhanson commented:
@cameronblandford @htpvu Starting to get harder for me to break. I was able to do some pretty deep testing. 

### Submission Details page
* Submitter field is still blank on Submission Details page. Ideally a Prepared By name should also be added. 

### Search user
* Can it search when I hit enter?
* Headings are left aligned everywhere else in the system - can we left align "Search users?" and the text under.
* Perhaps a light grey vertical divider between the two columns?
* Enhancement: Could the "remove submitter" link be closer to the selected submitter? or maybe stand out more like the grant remove button.

### Workflow, general
* Start a new submission, assign grants. Go back to basics page, pick a person to do the submission on behalf of - your grants selections are not cleared and so you can mix the two people's grants.
* I start a no frills submission (no doi, no grant) try to next on files without adding. It says please add. I add a file then next again, it won't let me proceed
* Do 2 submissions back to back, both on behalf of someone else. The file from the first submission is still listed in the second, and the comment is there too.

### Proxy submitter
* Submitter sends back for changes - preparer goes in to edit. When I get to JScholarship metadata page, it won't let me go past without signing the agreement - the agreement is hidden so I'm stuck.
* Submitter sends back for changes, preparer goes in to edit. Add a file. On the final submission review page before submitting changes for approval again, I can only see the file I just added, it does not show the files I added on the first update.
* Someone prepares a submission for me, I can cancel it without leaving a comment - is that intentional?
* Preparer prepares a submission with 1 NIH grant, submits for approval. Submitter goes in to Edit Submission instead of approving. They add second NIH grant, which does not change repo requirements. Next, next, the Details tab has no form.
![image](https://user-images.githubusercontent.com/4070836/47389909-fe780980-d6e3-11e8-9c26-857c34f90c8a.png)
* One to check with @htpvu about - When the repository is linked, the preparer is required to open the e.g. eric link but the submitter is not. I'm wondering if this should be switched in that the submitter must look at the linked repository, and the preparer does not have to.

### Submission using new user
* Do a submission using a new user, when you get to review page, there are some empty parentheses in the message above the comment box where there should be a name. "a notification will be sent to ()"
* Do a submission with a new user, the submissionEvent goes in with the eventType of "approval-requested" instead of "approval-requested-newuser"

@htpvu General thought: I think we may need to think about the `/submissions/` table a little more - perhaps add buttons in the Actions column to the ones the current user can act on. I was switching back and forth between 2 users, and it became confusing to figure out which ones I needed to do something with vs which one I was waiting for someone else.

cameronblandford commented:
- [ ] Submitter field is still blank on Submission Details page. Ideally a Prepared By name should also be added.
- [ ] Can it search when I hit enter?
- [ ] Headings are left aligned everywhere else in the system - can we left align "Search users?" and the text under.
- [ ] Perhaps a light grey vertical divider between the two columns?
- [ ] Enhancement: Could the "remove submitter" link be closer to the selected submitter? or maybe stand out more like the grant remove button.

- [ ] Start a new submission, assign grants. Go back to basics page, pick a person to do the submission on behalf of - your grants selections are not cleared and so you can mix the two people's grants.
- [ ] I start a no frills submission (no doi, no grant) try to next on files without adding. It says please add. I add a file then next again, it won't let me proceed
- [ ] Do 2 submissions back to back, both on behalf of someone else. The file from the first submission is still listed in the second, and the comment is there too.

- [ ] Submitter sends back for changes - preparer goes in to edit. When I get to JScholarship metadata page, it won't let me go past without signing the agreement - the agreement is hidden so I'm stuck.
- [ ] Submitter sends back for changes, preparer goes in to edit. Add a file. On the final submission review page before submitting changes for approval again, I can only see the file I just added, it does not show the files I added on the first update.
- [ ] Someone prepares a submission for me, I can cancel it without leaving a comment - is that intentional?
- [ ] Preparer prepares a submission with 1 NIH grant, submits for approval. Submitter goes in to Edit Submission instead of approving. They add second NIH grant, which does not change repo requirements. Next, next, the Details tab has no form.
- [ ] One to check with @htpvu about - When the repository is linked, the preparer is required to open the e.g. eric link but the submitter is not. I'm wondering if this should be switched in that the submitter must look at the linked repository, and the preparer does not have to.

- [ ] Do a submission using a new user, when you get to review page, there are some empty parentheses in the message above the comment box where there should be a name. "a notification will be sent to ()"
- [ ] Do a submission with a new user, the submissionEvent goes in with the eventType of "approval-requested" instead of "approval-requested-newuser"

karenhanson commented:
@cameronblandford I cannot create a new submission on the latest commit. I will investigate to try and get more information.
![image](https://user-images.githubusercontent.com/4070836/47585179-8c453600-d92a-11e8-874f-d1f80a7ca59a.png)
`/submissions/` and `/grants/` works fine

htpvu commented:
@karenhanson @cameronblandford I chatted to sayeed about the Blocked item relating to web linked repository. He wants us to have the requirement for clicking on weblinked repositories to be enforced on Submitter rather than preparer.  Please ping me if this causes complication.

karenhanson commented:
A handful of issues remain, but mostly this is working, so goal is to merge this PR and then log minor issues separately.
